File: technology-lab/java-showcase/src/main/java/no/ab/showcase/EventSummary.java
package no.ab.showcase;

import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/** Aggregated view over a list of events, computed with streams. */
public record EventSummary(int total, long totalBytes, Map<WorkEvent.Type, Long> byType) {

    public static EventSummary of(List<WorkEvent> events) {
        Map<WorkEvent.Type, Long> counted = new EnumMap<>(WorkEvent.Type.class);
        for (WorkEvent.Type type : WorkEvent.Type.values()) {
            counted.put(type, 0L);
        }
        counted.putAll(events.stream().collect(Collectors.groupingBy(WorkEvent::type, Collectors.counting())));

        long totalBytes = events.stream().mapToLong(WorkEvent::bytes).sum();
        return new EventSummary(events.size(), totalBytes, Map.copyOf(counted));
    }
}

File: technology-lab/kotlin-showcase/src/main/kotlin/no/ab/showcase/EventSummary.kt
package no.ab.showcase

/** Aggregated view over a list of events. */
data class EventSummary(val total: Int, val totalBytes: Long, val byType: Map<EventType, Int>) {
    companion object {
        fun of(events: List<WorkEvent>): EventSummary = EventSummary(
            total = events.size,
            totalBytes = events.sumOf { it.bytes.toLong() },
            byType = EventType.entries.associateWith { type -> events.count { it.type == type } },
        )
    }
}
